lib/TPP/Transforms/LoopShuffle.cpp

```diff
@@ -0,0 +1,109 @@
+//===- LoopShuffle.cpp -----------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file shuffles parallel loop based on user input
+//
+//===----------------------------------------------------------------------===//
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+#include "llvm/Support/Debug.h"
+#include <list>
+
+#define DEBUG_TYPE "loop-shuffle"
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DECL_LOOPSHUFFLEPASS
+#define GEN_PASS_DEF_LOOPSHUFFLEPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::scf;
+using namespace std;
+
+namespace mlir {
+namespace tpp {
+
+static LogicalResult loopShuffle(scf::ForallOp op,
+                                 ArrayRef<unsigned> shuffleOrder) {
+  OpBuilder b(op);
+  IRRewriter rewriter(b.getContext());
+  if (shuffleOrder.size() != op.getInductionVars().size())
+    return rewriter.notifyMatchFailure(op, "Number of indices incorrect");
+  for (size_t i = 0; i < op.getInductionVars().size(); i++) {
+    bool match = false;
+    for (size_t j = 0; j < shuffleOrder.size(); j++)
+      if (i == shuffleOrder[j]) {
+        match = true;
+        break;
+      }
+    if (!match) {
+      return rewriter.notifyMatchFailure(op, "Indices missed");
+    }
+  }
+
+  SmallVector<int64_t> lbs, ubs, steps;
+
+  for (size_t i = 0; i < op.getStaticLowerBound().size(); i++) {
```

Comment:
If lb/ub/step needs to have the same size as `shuffleOrder`, you can collapse all this in a single loop. 

If you use iterators, you can even use `llvm::zip_equals`.